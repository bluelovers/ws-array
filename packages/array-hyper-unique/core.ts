import { defaultFilter, equals, defaultChecker } from './lib/util';
import { IOptions } from './lib/types';

export * from './lib/types';
export * from './lib/util';

export function array_unique<T>(arr: T, options: IOptions<T> = {}): T
{
	if (!Array.isArray(arr))
	{
		throw new TypeError(`Expected an Array but got ${typeof arr}.`)
	}

	const cb = defaultFilter(options);

	if (options.overwrite)
	{
		let index = arr.length;

		while (index--)
		{
			let val = arr[index];

			if (!cb(val, index, arr))
			{
				arr.splice(index, 1);
			}
		}

		return arr;
	}

	// @ts-ignore
	return arr.filter(cb);
}

export function array_unique_overwrite<T>(arr: T, options: IOptions<T> = {}): T
{
	return array_unique(arr, {
		...options,
		overwrite: true,
	});
}

export function lazy_unique<T extends any[]>(arr: T): T
export function lazy_unique<T, T1, T2>(a1: T1, a2: T2, ...arr: T[]): Array<T | T1 | T2>
export function lazy_unique<T>(...arr: Array<T | T[]>): T | (T | T[])[]
// @ts-ignore
export function lazy_unique<T>(...arr: Array<T | T[]>)
{
	if (arr.length > 1)
	{
		return array_unique(arr);
	}

	return array_unique(arr[0]);
}

export function lazy_unique_overwrite<T>(...arr: Array<T | T[]>)
{
	if (arr.length > 1)
	{
		return array_unique_overwrite(arr);
	}

	return array_unique_overwrite(arr[0]);
}

lazy_unique.array_unique = array_unique;
lazy_unique.array_unique_overwrite = array_unique_overwrite;

lazy_unique.lazy_unique_overwrite = lazy_unique_overwrite;

lazy_unique.equals = equals;
lazy_unique.defaultFilter = defaultFilter;
lazy_unique.defaultChecker = defaultChecker;

lazy_unique.lazy_unique = lazy_unique;
lazy_unique.default = lazy_unique;

Object.defineProperty(lazy_unique, "__esModule", { value: true });

export default lazy_unique
