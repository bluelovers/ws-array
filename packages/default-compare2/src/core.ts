import { _typeOf } from './util';

export const enum EnumSortCompareResult
{
	KEEP = 0,
	RIGHT = 1,
	LEFT = -1,
}

export function numberCompare(a: unknown, b: unknown)
{
	if (a < b)
	{
		return EnumSortCompareResult.LEFT
	}
	else if (a > b)
	{
		return EnumSortCompareResult.RIGHT
	}

	return EnumSortCompareResult.KEEP
}

/**
 * will return undefined when a and b not null/undefined
 */
export function nullCompare(a: unknown, b: unknown)
{
	const typeA = _typeOf(a);
	const typeB = _typeOf(b);

	if (typeA === 'null')
	{
		return typeB === 'null' ? EnumSortCompareResult.KEEP : (typeB === 'undefined'
			? EnumSortCompareResult.LEFT
			: EnumSortCompareResult.RIGHT);
	}
	else if (typeA === 'undefined')
	{
		return typeB === 'null' ? EnumSortCompareResult.RIGHT : (typeB === 'undefined'
			? EnumSortCompareResult.KEEP
			: EnumSortCompareResult.RIGHT);
	}
	else if (typeB === 'null' || typeB === 'undefined')
	{
		return EnumSortCompareResult.LEFT;
	}
}

export function symbolCompare(a: unknown, b: unknown)
{
	const typeA = _typeOf(a);
	const typeB = _typeOf(b);

	if (typeA === 'symbol')
	{
		return typeB === 'symbol' ? numberCompare(String(a), String(b)) : EnumSortCompareResult.LEFT
	}
	else if (typeB === 'symbol')
	{
		return EnumSortCompareResult.RIGHT
	}

	return numberCompare(a, b)
}

export function dateCompare(a: Date, b: Date)
{
	return a.getTime() - b.getTime()
}

export function stringCompareLocale(a: string, b: string)
{
	return a.localeCompare(b)
}
