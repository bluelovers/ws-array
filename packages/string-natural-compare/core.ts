/**
 * Created by user on 2020/6/4.
 */

import _naturalCompare from 'string-natural-compare';

export interface IOptionsStringNaturalCompare
{
	/**
	 * Set to true to compare strings case-insensitively. Default: false.
	 */
	caseInsensitive?: boolean,
	/**
	 * A string of characters that define a custom character ordering. Default: undefined.
	 */
	alphabet?: string,
}

export interface IOptionsNaturalCompare extends IOptionsStringNaturalCompare
{
	desc?: boolean,
}

/**
 * Compare alphanumeric strings the same way a human would,
 * using a natural order algorithm
 * (originally known as the alphanum algorithm)
 * where numeric characters are sorted
 * based on their numeric values rather than their ASCII values.
 */
export function naturalCompare(a, b, opts?: IOptionsNaturalCompare)
{
	let i: number;

	if (typeof a === 'number' && typeof b === 'number')
	{
		i = a - b
	}
	else
	{
		if (typeof a === 'number')
		{
			a = String(a);
		}
		else if (typeof b === 'number')
		{
			b = String(b);
		}

		if (a === b)
		{
			return 0
		}

		i = _naturalCompare(a, b, opts)
	}

	if (i !== 0 && opts?.desc)
	{
		i = 0 - i;
	}

	return i
}

/**
 * create compare with preset options
 */
export function createNew(opts?: IOptionsNaturalCompare)
{
	return (a, b) => naturalCompare(a, b, opts)
}

/**
 * compare strings case-insensitively
 */
export const compareCaseInsensitive = createNew({
	caseInsensitive: true,
})

naturalCompare.createNew = createNew;
naturalCompare.compareCaseInsensitive = compareCaseInsensitive;
naturalCompare.default = naturalCompare;

export default naturalCompare
