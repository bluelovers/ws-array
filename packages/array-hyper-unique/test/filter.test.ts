/**
 * Created by user on 2018/5/30/030.
 */

import getTestcase from './_data';
import { array_unique, IOptions } from '../core';

describe(`(v) => v`, () =>
{
	let options: IOptions<any> = {
		filter(v)
		{
			return v;
		},
	};

	getTestcase()
		.forEach(function ({
			data,
			expected,
			label,
		})
		{
			expected = expected.filter(options.filter);

			_testIt(data, expected, label, options)
		})
	;

	function _testIt(data, expected, label: string = `label`, options?)
	{
		it(label, () =>
		{
			//console.log('it:inner', currentTest.title);
			//console.log('it:inner', currentTest.fullTitle());

			let len = data.length;

			let actual = array_unique(data, options);

			expect(actual).toBeTruthy();
			expect(actual).toEqual(expected);

			expect(actual).not.toEqual(data);
			expect(actual.length).not.toEqual(len);

			expect(actual).toMatchSnapshot();
			expect(data).toMatchSnapshot();

		});
	}
});

