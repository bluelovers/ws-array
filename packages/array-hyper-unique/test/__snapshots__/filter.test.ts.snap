// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(v) => v ArrayBuffer 1`] = `
Array [
  ArrayBuffer [],
  ArrayBuffer [],
]
`;

exports[`(v) => v ArrayBuffer 2`] = `
Array [
  ArrayBuffer [],
  ArrayBuffer [],
  ArrayBuffer [],
]
`;

exports[`(v) => v Buffer & ArrayBuffer 1`] = `
Array [
  Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  Array [
    0,
    0,
    0,
    0,
    0,
  ],
  Object {},
  ArrayBuffer [],
]
`;

exports[`(v) => v Buffer & ArrayBuffer 2`] = `
Array [
  Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  Array [
    0,
    0,
    0,
    0,
    0,
  ],
  Object {},
  Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  ArrayBuffer [],
  ArrayBuffer [],
  Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  Array [
    0,
    0,
    0,
    0,
    0,
  ],
]
`;

exports[`(v) => v Buffer 1`] = `
Array [
  Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  Array [
    0,
    0,
    0,
    0,
    0,
  ],
  Object {},
]
`;

exports[`(v) => v Buffer 2`] = `
Array [
  Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  Array [
    0,
    0,
    0,
    0,
    0,
  ],
  Object {},
  Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "type": "Buffer",
  },
  Array [
    0,
    0,
    0,
    0,
    0,
  ],
]
`;

exports[`(v) => v Map 1`] = `
Array [
  Map {
    "a" => 1,
    "b" => 2,
    "c" => 3,
  },
  Object {
    "a": 1,
    "b": 2,
    "c": 3,
  },
]
`;

exports[`(v) => v Map 2`] = `
Array [
  Map {
    "a" => 1,
    "b" => 2,
    "c" => 3,
  },
  Map {
    "c" => 3,
    "b" => 2,
    "a" => 1,
  },
  Object {
    "a": 1,
    "b": 2,
    "c": 3,
  },
  Object {
    "a": 1,
    "b": 2,
    "c": 3,
  },
]
`;

exports[`(v) => v RegExp 1`] = `
Array [
  /1/,
  Object {},
  /1/g,
  /1/i,
]
`;

exports[`(v) => v RegExp 2`] = `
Array [
  /1/,
  Object {},
  /1/g,
  /1/i,
  /1/,
  Object {},
  /1/g,
  /1/i,
]
`;

exports[`(v) => v boolean 1`] = `
Array [
  true,
]
`;

exports[`(v) => v boolean 2`] = `
Array [
  true,
  true,
  false,
  false,
  true,
]
`;

exports[`(v) => v boolean 2 1`] = `
Array [
  true,
  Infinity,
]
`;

exports[`(v) => v boolean 2 2`] = `
Array [
  true,
  true,
  false,
  false,
  true,
  null,
  0,
  undefined,
  null,
  0,
  undefined,
  0,
  NaN,
  Infinity,
  "",
]
`;

exports[`(v) => v function 1`] = `
Array [
  [Function],
  [Function],
  [Function],
]
`;

exports[`(v) => v function 2`] = `
Array [
  [Function],
  [Function],
  [Function],
  [Function],
  [Function],
  [Function],
]
`;

exports[`(v) => v function 2 1`] = `
Array [
  [Function],
  [Function],
  [Function],
  Object {},
]
`;

exports[`(v) => v function 2 2`] = `
Array [
  [Function],
  [Function],
  [Function],
  [Function],
  [Function],
  Object {},
  [Function],
  Object {},
]
`;

exports[`(v) => v main mixin test 1`] = `
Array [
  1,
  true,
  Array [
    "a",
    "b",
    "c",
  ],
  Array [
    "a",
    "c",
    "b",
  ],
  Object {
    "a": Object {
      "b": 2,
    },
  },
  Object {
    "a": Object {
      "b": 3,
    },
  },
  Object {
    "a": Object {
      "b": undefined,
    },
  },
  Object {
    "a": Object {},
  },
  Object {
    "a": Object {
      "b": 3,
      "c": undefined,
    },
  },
  /1/,
  Object {},
  /1/g,
  /1/i,
  "1",
  "0",
  "true",
  "undefined",
  "null",
  "false",
  Infinity,
  [Function],
  [Function],
  [Function],
]
`;

exports[`(v) => v main mixin test 2`] = `
Array [
  1,
  0,
  true,
  undefined,
  null,
  false,
  Array [
    "a",
    "b",
    "c",
  ],
  Array [
    "a",
    "b",
    "c",
  ],
  Array [
    "a",
    "c",
    "b",
  ],
  Object {
    "a": Object {
      "b": 2,
    },
  },
  Object {
    "a": Object {
      "b": 2,
    },
  },
  Object {
    "a": Object {
      "b": 3,
    },
  },
  Object {
    "a": Object {
      "b": undefined,
    },
  },
  Object {
    "a": Object {},
  },
  Object {
    "a": Object {
      "b": 3,
      "c": undefined,
    },
  },
  /1/,
  Object {},
  /1/,
  /1/g,
  /1/i,
  Object {},
  1,
  0,
  true,
  undefined,
  null,
  false,
  Array [
    "a",
    "b",
    "c",
  ],
  Array [
    "a",
    "b",
    "c",
  ],
  Array [
    "a",
    "c",
    "b",
  ],
  Object {
    "a": Object {
      "b": 2,
    },
  },
  Object {
    "a": Object {
      "b": 2,
    },
  },
  Object {
    "a": Object {
      "b": 3,
    },
  },
  Object {
    "a": Object {
      "b": undefined,
    },
  },
  Object {
    "a": Object {},
  },
  Object {
    "a": Object {
      "b": 3,
      "c": undefined,
    },
  },
  /1/,
  Object {},
  /1/,
  /1/g,
  /1/i,
  Object {},
  "1",
  "0",
  "true",
  "undefined",
  "null",
  "false",
  "",
  NaN,
  NaN,
  NaN,
  Infinity,
  [Function],
  [Function],
  [Function],
  [Function],
  [Function],
  [Function],
]
`;

exports[`(v) => v number 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`(v) => v number 2`] = `
Array [
  1,
  2,
  3,
  3,
  0,
]
`;

exports[`(v) => v number 2 1`] = `
Array [
  1,
  2,
  3,
  Infinity,
]
`;

exports[`(v) => v number 2 2`] = `
Array [
  1,
  2,
  3,
  3,
  0,
  NaN,
  NaN,
  NaN,
  Infinity,
]
`;

exports[`(v) => v object 1`] = `
Array [
  Object {
    "a": Object {
      "b": 2,
    },
  },
  Object {
    "a": Object {
      "b": 3,
    },
  },
]
`;

exports[`(v) => v object 2`] = `
Array [
  Object {
    "a": Object {
      "b": 2,
    },
  },
  Object {
    "a": Object {
      "b": 2,
    },
  },
  Object {
    "a": Object {
      "b": 3,
    },
  },
]
`;

exports[`(v) => v string 1`] = `
Array [
  "1",
  "2",
  "3",
  "0",
  "a",
  "b",
  "c",
  "d",
]
`;

exports[`(v) => v string 2`] = `
Array [
  "1",
  "2",
  "3",
  "3",
  "0",
  "",
  "a",
  "b",
  "c",
  "a",
  "b",
  "d",
]
`;

exports[`(v) => v string[] 1`] = `
Array [
  Array [
    "a",
    "b",
    "c",
  ],
]
`;

exports[`(v) => v string[] 2`] = `
Array [
  Array [
    "a",
    "b",
    "c",
  ],
  Array [
    "a",
    "b",
    "c",
  ],
]
`;
